File: algorithms/insertion_sort_list.rb
# https://leetcode.com/problems/insertion-sort-list/
#
# Sort a linked list using insertion sort.


# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end


# @param {ListNode} head
# @return {ListNode}
def insertion_sort_list(head)
  return head if head.nil? || head.next.nil?

  dummy = ListNode.new(0)
  prev, curr, last = dummy, head, head

  while curr
    last = curr.next

    while prev.next && prev.next.val < curr.val
      prev = prev.next
    end

    curr.next = prev.next
    prev.next = curr
    curr = last
    prev = dummy
  end

  dummy.next
end

File: algorithms/insertion_sort_list.java
/**
 * https://leetcode.com/problems/insertion-sort-list/
 * Sort a linked list using insertion sort.
 *
 *
 *
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
	public ListNode insertionSortList(ListNode head) {
		if (head == null || head.next == null) return head;

		ListNode dummy = new ListNode(0);
		ListNode curt = head, prev = dummy, next = head;
		while (curt != null) {
			next = curt.next;

			while (prev.next != null && prev.next.val < curt.val) {
				prev = prev.next;
			}

			curt.next = prev.next;
			prev.next = curt;
			curt = next;
			prev = dummy;
		}

		return dummy.next;
	}
}
